### PR TITLE
Bypass vector copy in MatrixFree::cell_loop with MPI

### PR DESCRIPTION
**Pull request by kronbichler**
The CUDA code with MPI would previously copy the vector in `CUDAWrappers::MatrixFree::cell_loop` for handling the ghost entries. However, if we have compatible partitioners, we can bypass that and simply let the vectors call `update_ghost_values()` and `compress()`.

### CODE REVIEW

**Comment by kronbichler** [comment]
FYI @peterrum

**Comment by kronbichler** [comment]
When running with multiple GPUs the CG solver in step-64 does not converge any more - I have to investigate what happens, so putting the `do no merge` label on for now.

File: examples/step-64/step-64.cu [hunk 1]
```diff
@@ -304,6 +307,15 @@ namespace Step64
   }
 
 
+
+  template <int dim, int fe_degree>
+  void HelmholtzOperator<dim, fe_degree>::initialize_dof_vector(
+    LinearAlgebra::distributed::Vector<double, MemorySpace::CUDA> &vec) const
+  {
+    mf_data.initialize_dof_vector(vec);
```
Comment (peterrum) [inline examples/step-64/step-64.cu]: As far as I see it, this function does not zero out the entries. Inside it calls `reinit(const std::shared_ptr<const Utilities::MPI::Partitioner> &partitioner)`, which does not  have the flag `const bool omit_zeroing_entries = false`. I do not know if this helps. Everything else looks good (from a standpoint that I do not have access to a compiler).

Comment (Rombur) [reply examples/step-64/step-64.cu]: This might be a problem. All the constrained entries need to be zero out.

Comment (kronbichler) [reply examples/step-64/step-64.cu]: The [function `reinit(Partitioner)`](https://github.com/dealii/dealii/blob/master/include/deal.II/lac/la_parallel_vector.templates.h#L511-L536) should indeed zero the vector, including the ghosts, as it calls `operator=(Number())`. I also checked that the `zero_out_ghosts()` function is doing the right thing.

**Comment by kronbichler** [comment]
Note that I had `DEAL_II_MPI_WITH_CUDA_SUPPORT=ON` and it is unclear whether the RDMA works correctly (e.g. some synchronization missing in our code or some bug on the machine). When I disable this part and go through the host, my program runs fine (I tested on 2 and 5 GPUs). Thus, I'm removing the WIP flags; this should be ready for review.

**Comment by kronbichler** [comment]
/rebuild

(This should only trigger on the CUDA machines and thus not the tester, but let's play safe given that are non-trivial changes.)

File: doc/news/changes/minor/20190927MartinKronbichler [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Improved: CUDAWrappers::MatrixFree::cell_loop() can now detect whether the
+given vectors have the same parallel MPI partitioners as the matrix-free
```
Comment (masterleinad) [inline doc/news/changes/minor/20190927MartinKronbichler]:
```suggestion
given vectors have the same parallel MPI partitioner as the matrix-free
```

**Comment by masterleinad** [comment]
I don't see more tests failing enabling CUDA-aware MPI support compared to not doing it. We need to investigate that separately.

**Comment by kronbichler** [comment]
This likely depends on the MPI implementation. On the machine I'm testing on (Piz Daint), I have seen that to fix this we need
```
diff --git a/include/deal.II/base/partitioner.templates.h b/include/deal.II/base/partitioner.templates.h
index 167e01c45c..da662e10de 100644
--- a/include/deal.II/base/partitioner.templates.h
+++ b/include/deal.II/base/partitioner.templates.h
@@ -126,6 +126,7 @@ namespace Utilities
                   import_indices_plain_dev[i].first.get(),
                   locally_owned_array.data(),
                   chunk_size);
+              cudaDeviceSynchronize();
             }
           else
 #    endif
@@ -400,6 +401,11 @@ namespace Utilities
             ExcMessage("Index overflow: Maximum message size in MPI is 2GB. "
                        "The number of ghost entries times the size of 'Number' "
                        "exceeds this value. This is not supported."));
+#    if defined(DEAL_II_COMPILER_CUDA_AWARE) && \
+      defined(DEAL_II_MPI_WITH_CUDA_SUPPORT)
+          if (std::is_same<MemorySpaceType, MemorySpace::CUDA>::value)
+            cudaDeviceSynchronize();
+#    endif
           const int ierr =
             MPI_Isend(ghost_array_ptr,
                       ghost_targets_data[i].second * sizeof(Number),
```
I'm hesitating to add this now unless someone else confirms that this is necessary and it's not a "bug" in the MPI library. In my case, this is necessary because `MPI_Isend` does not wait for the CUDA kernels to finish, so the data to be exchanged might not be ready at the time MPI copies the data from one rank to the other. In any case, we have a solution to the problem.

**Comment by masterleinad** [comment]
I agree with these changes. I also get different numbers of iterations in `step-64` with CUDA-aware MPI.

If you already have the changes in branch please turn them into a pull request. Otherwise, I can do that as well.

**Pull request merged**